backend/src/main/java/fk/prof/backend/worker/BackendDaemon.java

```diff
@@ -0,0 +1,198 @@
+
+    this.leaderReadContext = leaderReadContext;
+    this.processGroupAssociationStore = processGroupAssociationStore;
+    this.aggregationWindowLookupStore = aggregationWindowLookupStore;
+    this.simultaneousWorkAssignmentCounter = simultaneousWorkAssignmentCounter;
+  }
+
+  @Override
+  public void start() {
+    httpClient = buildHttpClient();
+    aggregationWindowPlannerStore = buildAggregationWindowPlannerStore();
+    postLoadToLeader();
+  }
+
+  private ProfHttpClient buildHttpClient() {
+    JsonObject httpClientConfig = configManager.getHttpClientConfig();
+    ProfHttpClient httpClient = ProfHttpClient.newBuilder()
+        .keepAlive(httpClientConfig.getBoolean("keepalive", false))
+        .useCompression(httpClientConfig.getBoolean("compression", true))
+        .setConnectTimeoutInMs(httpClientConfig.getInteger("connect.timeout.ms", 2000))
+        .setIdleTimeoutInSeconds(httpClientConfig.getInteger("idle.timeout.secs", 3))
+        .setMaxAttempts(httpClientConfig.getInteger("max.attempts", 1))
+        .build(vertx);
+    return httpClient;
+  }
+
+  private AggregationWindowPlannerStore buildAggregationWindowPlannerStore() {
+    return new AggregationWindowPlannerStore(
+        vertx,
+        config().getInteger("aggregation.window.duration.mins", 30),
+        config().getInteger("aggregation.window.end.tolerance.secs", 120),
+        config().getInteger("workprofile.refresh.offset.secs", 300),
+        config().getInteger("scheduling.buffer.secs", 30),
+        config().getInteger("work.assignment.max.delay.secs", 120),
+        simultaneousWorkAssignmentCounter,
+        aggregationWindowLookupStore,
+        this::getWorkFromLeader);
+  }
+
+  private void postLoadToLeader() {
+    String leaderIPAddress;
+    if((leaderIPAddress = leaderReadContext.getLeaderIPAddress()) != null) {
+
+      //TODO: load = 0.5 hard-coded right now. Replace this with dynamic load computation in future
+      float load = 0.5f;
+
+      try {
+        httpClient.requestAsync(
+            HttpMethod.POST,
+            leaderIPAddress,
+            leaderHttpPort,
+            ApiPathConstants.LEADER_POST_LOAD,
+            ProtoUtil.buildBufferFromProto(
+                BackendDTO.LoadReportRequest.newBuilder()
+                    .setIp(ipAddress)
+                    .setPort(backendHttpPort)
+                    .setLoad(load)
+                    .setCurrTick(loadTickCounter++)
+                    .build()))
+            .setHandler(ar -> {
+              if(ar.succeeded()) {
+                if(ar.result().getStatusCode() == 200) {
+                  try {
+                    Recorder.ProcessGroups assignedProcessGroups = ProtoUtil.buildProtoFromBuffer(Recorder.ProcessGroups.parser(), ar.result().getResponse());
+                    processGroupAssociationStore.updateProcessGroupAssociations(assignedProcessGroups, (processGroupDetail, processGroupAssociationResult) -> {
```

Comment:
Assignment of PG on leader is lazy and only in response to /association request made by some recorder. Recorder will make such a request if it cannot talk to backend in the first place, resulting in reassignment and setup of new aggregation window(with a different start time) on another backend machine. Since serialized files have start time in the name, we should not be worried about possibility of duplicate files being written (or competing for writes) in d42.

The scenario you are talking about can happen when n/w partition occurs and some recorders(let's call that set S1) can talk to previous backend(B1), others cannot(S2). Recorders in S2, then call association and get a new backend assigned to PG(B2) if leader determines B1 is defunct. If B1 had recovered before that or healthy regardless according to leader, reassignment does not happen, and leader still returns B1.
In case of leader returning B2 because according to leader b1 is defunct, recorders in S1 are still talking to B1. That is a problem yes, but only for the current aggregation window. I can add a check in /leader/work API to return work assignment only if B1 is still assigned with the PG. In this case two files are written to disk but as I said with different start times, so not a problem. In the case where n/w recovers before aggregation window end, backend is able to report load to leader, finds out it is not associated with PG anymore and expires self aggregation window.

So if we tweak /leader/work API such that backend sends self details as well in the request and leader checks for association before handing out work, above should not be a issue.